ddtrace/contrib/asgi/middleware.py

```diff
@@ -0,0 +1,113 @@
+from ddtrace import tracer as global_tracer
+from ddtrace.constants import ANALYTICS_SAMPLE_RATE_KEY
+from ddtrace.ext import SpanTypes, http
+from ddtrace.http import store_request_headers, store_response_headers
+from ddtrace.propagation.http import HTTPPropagator
+from ddtrace.settings import config
+from ddtrace.vendor.six.moves import urllib
+
+from ...internal.logger import get_logger
+from .constants import ASGI_SPEC_VERSION, ASGI_VERSION
+from .utils import guarantee_single_callable
+
+log = get_logger(__name__)
+
+config._add("asgi", dict(service_name=config._get_service(default="asgi"), distributed_tracing=True,))
+
+
+def _extract_tags_from_scope(scope):
+    tags = {}
+
+    http_method = scope.get("method")
+    if http_method:
+        tags[http.METHOD] = http_method
+
+    query_string = None
+    if config.asgi.trace_query_string:
+        query_string = scope.get("query_string")
+        if isinstance(query_string, bytes):
+            query_string = query_string.decode()
+        tags[http.QUERY_STRING] = query_string
+
+    server = scope.get("server")
+    if server and len(server) == 2:
+        port = server[1]
+        server_host = server[0] + (":" + str(port) if port != 80 else "")
+        full_path = scope.get("root_path", "") + scope.get("path", "")
+        http_url = scope.get("scheme", "http") + "://" + server_host + full_path
+        if query_string:
+            http_url = http_url + ("?" + urllib.parse.unquote(query_string))
+        tags[http.URL] = http_url
+
+    http_version = scope.get("http_version")
+    if http_version:
+        tags[http.VERSION] = http_version
+
+    scope_asgi = scope.get("asgi")
+
+    if scope_asgi and "version" in scope_asgi:
+        tags[ASGI_VERSION] = scope_asgi["version"]
+
+    if scope_asgi and "version" in scope_asgi:
+        tags[ASGI_SPEC_VERSION] = scope_asgi["spec_version"]
+
+    return tags
+
+
+class TraceMiddleware:
+    """
+    ASGI application middleware that traces the requests.
+
+    Args:
+        app: The ASGI application.
+        tracer: Custom tracer. Defaults to the global tracer.
+    """
+
+    def __init__(self, app, tracer=global_tracer):
+        self.app = guarantee_single_callable(app)
+        self.tracer = tracer
+
+    async def __call__(self, scope, receive, send):
+        if scope["type"] != "http":
+            return await self.app(scope, receive, send)
+
+        headers = {k: v for (k, v) in scope["headers"]}
+
+        if config.asgi.distributed_tracing:
+            propagator = HTTPPropagator()
+            context = propagator.extract(headers)
+            if context.trace_id:
+                self.tracer.context_provider.activate(context)
+
+        resource = "{} {}".format(scope["method"], scope["path"])
+
+        span = self.tracer.trace(
```

Comment:
Hmm this might be a problem. I don't know enough about the internals of how asgi works but we might have a context propagation problem here in Python 3.5/3.6 (Pythons without ContextVars). Do we have a test case either in the suite or manual that verifies that if two requests come in concurrently that the propagation works correctly? This might be tricky to test.

For example say there are requests r1, r2 that come in simultaneously. r1 calls `tracer.trace()` which activates a span in the current context (for Python 3.5/3.6 the context is the thread). r2 then comes in before r1 finishes and calls `tracer.trace()` but r1's span is active and so r2's span will be a child of r1's span.

It shouldn't be an issue in Python 3.7 and above because the asyncio supports context vars as of Python 3.7, but I suspect it will be for 3.5, 3.6.

If this is an issue we may have to make sure the tracer is configured to use the asyncio context provider that we have.

Reply:
I've now added documentation and tests for py36

Reply:
challenge with py35 is that httpx (which we are using for testing multiple async requests) requires Python 3.6+